pkg/destroy/gcp/network.go

```diff
@@ -0,0 +1,647 @@
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	req := o.computeSvc.BackendServices.List(o.ProjectID).Fields(googleapi.Field(fields))
+	if len(filter) > 0 {
+		req = req.Filter(filter)
+	}
+	err := req.Pages(ctx, func(list *compute.BackendServiceList) error {
+		for _, backendService := range list.Items {
+			if filterFunc == nil || filterFunc != nil && filterFunc(backendService) {
+				o.Logger.Debugf("Found backend service: %s", backendService.Name)
+				result = append(result, backendService.Name)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list backend services")
+	}
+	return result, nil
+}
+
+func (o *ClusterUninstaller) deleteBackendService(name string) error {
+	o.Logger.Debugf("Deleting backend service %s", name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	_, err := o.computeSvc.BackendServices.Delete(o.ProjectID, name).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		return errors.Wrapf(err, "failed to delete backend service %s", name)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) destroyBackendServices() error {
+	backendServices, err := o.listClusterBackendServices()
+	if err != nil {
+		return err
+	}
+	for _, backendService := range backendServices {
+		err = o.deleteBackendService(backendService)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) listHealthChecks() ([]string, error) {
+	o.Logger.Debugf("Listing health checks")
+	result := []string{}
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	req := o.computeSvc.HealthChecks.List(o.ProjectID).Fields("items(name)").Filter(o.clusterIDFilter())
+	err := req.Pages(ctx, func(list *compute.HealthCheckList) error {
+		for _, healthCheck := range list.Items {
+			o.Logger.Debugf("Found health check: %s", healthCheck.Name)
+			result = append(result, healthCheck.Name)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list health checks")
+	}
+	return result, nil
+}
+
+func (o *ClusterUninstaller) deleteHealthCheck(name string) error {
+	o.Logger.Debugf("Deleting health check %s", name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	_, err := o.computeSvc.HealthChecks.Delete(o.ProjectID, name).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		return errors.Wrapf(err, "failed to delete health check %s", name)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) destroyHealthChecks() error {
+	healthChecks, err := o.listHealthChecks()
+	if err != nil {
+		return err
+	}
+	for _, healthCheck := range healthChecks {
+		err = o.deleteHealthCheck(healthCheck)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) listClusterTargetPools() ([]string, error) {
+	return o.listTargetPools("items(name)", o.clusterIDFilter(), nil)
+}
+
+func (o *ClusterUninstaller) listTargetPools(field string, filter string, filterFunc func(*compute.TargetPool) bool) ([]string, error) {
```

Comment:
can you add comments in terms of what is filter and filterFunc and how that affects the results.